docs/man/man3/AELogParser.3
.TH "AELogParser" 3 "Mon Mar 18 2024 18:44:24" "Version v0.0.8.5a" "ArtyK's Console Engine" \" -*- nroff -*-
.ad l
.nh
.SH NAME
AELogParser \- ArtyK's Engine's Log Parser; it parses the log files that \fBAELogger\fP writes\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fR#include <AELogParser\&.hpp>\fP
.PP
Inherits \fB__AEModuleBase< AELogParser >\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBAELogParser\fP (void)"
.br
.RI "Class constructor -- just initialises the object\&. "
.ti -1c
.RI "\fBAELogParser\fP (const std::string_view fname)"
.br
.RI "Class constructor -- opens the file and start the indexing process\&. "
.ti -1c
.RI "\fB~AELogParser\fP (void)"
.br
.RI "Class destructor -- closes the log file\&. "
.ti -1c
.RI "\fBcint\fP \fBopenLog\fP (const std::string_view fname)"
.br
.RI "Opens the log file, reading and parsing every single entry, classifying it by type\&. "
.ti -1c
.RI "\fBcint\fP \fBcloseLog\fP (void)"
.br
.RI "Closes the log file and clears the memory\&. "
.ti -1c
.RI "\fBcint\fP \fBnextEntry\fP (\fBAELogEntry\fP &entry, const \fBcint\fP severity=\fBAELOG_TYPE_DEBUG\fP, const bool strictSeverity=false, const std::string_view mname='')"
.br
.RI "Read the next \fIvalid\fP entry in the log file of the given severity and module name filter, and parse it to the given \fBAELogEntry\fP object\&. "
.ti -1c
.RI "\fBcint\fP \fBnextEntryName\fP (\fBAELogEntry\fP &entry, const std::string_view mname)"
.br
.RI "Read the next \fIvalid\fP entry in the log file of the given module name, and parse it into the given \fBAELogEntry\fP object\&. "
.ti -1c
.RI "\fBcint\fP \fBnextEntryType\fP (\fBAELogEntry\fP &entry, const \fBcint\fP severity, const bool strictSeverity=false)"
.br
.RI "Read the next \fIvalid\fP entry in the log file of the given severity, and parse it to the given \fBAELogEntry\fP object\&. "
.ti -1c
.RI "\fBllint\fP \fBnextEntryCursor\fP (const \fBcint\fP severity=\fBAELOG_TYPE_DEBUG\fP, const bool strictSeverity=false, const std::string_view mname='')"
.br
.RI "Read the next indexed \fIvalid\fP log entry in the file of the given severity and module name filter, and return it's file cursors in the file\&. "
.ti -1c
.RI "\fBllint\fP \fBnextEntryCursorName\fP (const std::string_view mname)"
.br
.RI "Read the next indexed \fIvalid\fP log entry in the file of the given module name filter, and return it's file cursors in the file\&. "
.ti -1c
.RI "\fBllint\fP \fBnextEntryCursorType\fP (const \fBcint\fP severity, const bool strictSeverity=false)"
.br
.RI "Read the next indexed \fIvalid\fP log entry in the file of the given severity filter and return it's file cursors in the file\&. "
.ti -1c
.RI "\fBllint\fP \fBcurrentEntryCursor\fP (void) const noexcept"
.br
.RI "Get the file cursors of the current valid entry This entry is the one that was read by a previous call to \fBAELogParser::nextEntry()\fP or AELogParser::getNextEntryIndex() "
.ti -1c
.RI "std::size_t \fBgetCurrentEntryIndex\fP (void) const noexcept"
.br
.RI "Return the order number (index) of the current log entry (that was attempted to be read) "
.ti -1c
.RI "\fBcint\fP \fBsetNextEntryIndex\fP (const std::size_t entryIndex) noexcept"
.br
.RI "Set the order number (index) of the next log entry to be parsed\&. "
.ti -1c
.RI "std::vector< \fBAELogEntryInfo\fP > \fBgetEntryCursorAll\fP (void) const noexcept"
.br
.RI "Get the list of file cursors of the valid entries, separated by type\&. "
.ti -1c
.RI "std::vector< \fBllint\fP > \fBgetInvalidEntryCursorAll\fP (void) const noexcept"
.br
.RI "Get the list of file cursors of the invalid entries\&. "
.ti -1c
.RI "\fBcint\fP \fBlogToQueue\fP (\fBAELogEntry\fP *&begin, const \fBcint\fP severity=\fBAELOG_TYPE_DEBUG\fP, const bool strictSeverity=false, const std::string_view mname='')"
.br
.RI "Same as the \fBAELogParser::nextEntry()\fP but on massive scale -- scans the whole log file and parses it to the freshly-allocated queue\&. "
.ti -1c
.RI "\fBcint\fP \fBlogToQueueType\fP (\fBAELogEntry\fP *&begin, const \fBcint\fP severity, const bool strictSeverity=false)"
.br
.RI "A shortcut to \fBAELogParser::logToQueue()\fP to filter only by type/severity of the log entries\&. "
.ti -1c
.RI "\fBcint\fP \fBlogToQueueName\fP (\fBAELogEntry\fP *&begin, const std::string_view mname)"
.br
.RI "A shortcut to \fBAELogParser::logToQueue()\fP to filter only by the module name of the log entries\&. "
.ti -1c
.RI "\fBullint\fP \fBamountEntriesValid\fP (const \fBcint\fP severity=\fBAELOG_TYPE_DEBUG\fP) const noexcept"
.br
.RI "Get the amount of valid entries in the log (with optional lowest severity setting)\&. "
.ti -1c
.RI "\fBullint\fP \fBamountEntriesInvalid\fP (void) const noexcept"
.br
.RI "Get the amount of invalid entries in the log\&. "
.ti -1c
.RI "\fBullint\fP \fBamountEntriesType\fP (const \fBcint\fP severity) const noexcept"
.br
.RI "Get the amount of entries of a certain type/severity\&. "
.ti -1c
.RI "\fBullint\fP \fBamountEntriesName\fP (const std::string_view mname) const noexcept"
.br
.RI "Get the amount of entries of a certain module name\&. "
.ti -1c
.RI "bool \fBcontainsSeverity\fP (const \fBcint\fP severity, const bool strictSeverity=false) const noexcept"
.br
.RI "Checks if the current log file contains the given severity (or higher, depending on the 'strict-severity' flag) "
.ti -1c
.RI "bool \fBisOpen\fP (void) const noexcept"
.br
.RI "Checks if the file is opened by the \fBAELogParser\fP\&. "
.ti -1c
.RI "bool \fBisClosed\fP (void) const noexcept"
.br
.RI "Checks if the file is closed by the \fBAELogParser\fP\&. "
.ti -1c
.RI "std::string \fBgetFullFileName\fP (void) const noexcept"
.br
.RI "Returns the full name of currently open file, including the relative path, if such was provided\&. "
.ti -1c
.RI "std::string \fBgetRelativePath\fP (void) const"
.br
.RI "Returns the relative path of the opened file\&. "
.ti -1c
.RI "std::string \fBgetFileName\fP (void) const"
.br
.RI "Returns the name of the opened file, devoid of the path\&. "
.ti -1c
.RI "std::string \fBgetFullPath\fP (void) const"
.br
.RI "Returns the full, absolute path of the opened file\&. "
.ti -1c
.RI "\fBAEFileReader\fP & \fBgetFileReader\fP (void) noexcept"
.br
.RI "Return the reference to the \fBAEFileReader\fP object of this instance of \fBAELogParser\fP\&. "
.ti -1c
.RI "constexpr const \fBullint\fP \fBgetModuleID\fP (void) noexcept"
.br
.RI "Get the \fBid number of the instance\fP of the module\&. "
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static \fBcint\fP \fBfilterQueue\fP (\fBAELogEntry\fP *&ptr, const \fBcint\fP severity, const bool strictSeverity, const std::string_view mname, const bool reallocQueue=true)"
.br
.RI "Filters the passed queue according to the given severity and modulename filters\&. "
.ti -1c
.RI "static \fBcint\fP \fBfilterQueueType\fP (\fBAELogEntry\fP *&ptr, const \fBcint\fP severity, const bool strictSeverity=false, const bool reallocQueue=true)"
.br
.RI "Filters the passed queue according to the given severity\&. "
.ti -1c
.RI "static \fBcint\fP \fBfilterQueueName\fP (\fBAELogEntry\fP *&ptr, const std::string_view mname, const bool reallocQueue=true)"
.br
.RI "Filters the passed queue according to the given modulename\&. "
.ti -1c
.RI "static \fBullint\fP \fBgetModuleAmount\fP (void) noexcept"
.br
.RI "Get the \fBamount of the currently-alive instances of module T\fP\&. "
.ti -1c
.RI "static constexpr const std::string_view \fBgetModuleName\fP (void) noexcept"
.br
.RI "Get the \fBname of the module\fP\&. "
.ti -1c
.RI "static constexpr const \fBullint\fP \fBgetInstanceCounter\fP (void) noexcept"
.br
.RI "Get the current value of the \fBinstance counter\fP of the module\&. "
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "const \fBullint\fP \fBm_ullModuleId\fP"
.br
.RI "The ID number of the current module\&. "
.in -1c
.SS "Static Protected Attributes"

.in +1c
.ti -1c
.RI "static std::atomic< std::size_t > \fBm_ullModulesAlive\fP"
.br
.RI "Amount of the currently-alive (not destroyed) instances of the module\&. "
.ti -1c
.RI "static std::atomic< \fBullint\fP > \fBm_ullInstanceCounter\fP"
.br
.RI "Amount of the instances of the module ever created during the program's runtime\&. "
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "\fBAELogEntryInfo\fP \fBfindNextEntry\fP (const \fBcint\fP severity, const std::string_view mname, const bool strictSeverity)"
.br
.RI "Finds the \fBAELogEntryInfo\fP corresponding to the next log entry with the specified severity and module name\&. "
.ti -1c
.RI "\fBREGISTER_MODULE\fP (\fBAELogParser\fP)"
.br
.in -1c
.SS "Static Private Member Functions"

.in +1c
.ti -1c
.RI "static constexpr \fBcint\fP \fBerrorFromAELEI\fP (const \fBAELogEntryInfo\fP &leInfo) noexcept"
.br
.RI "Determines the error codes from the invalid values in the passed \fBAELogEntryInfo\fP instance\&. "
.ti -1c
.RI "static constexpr bool \fBcheckSeverity\fP (const \fBcint\fP entrySeverity, const \fBcint\fP filterSeverity, const bool strictSeverity=false) noexcept"
.br
.RI "Checks if the passed severity is valid compared to the filtering severity\&. "
.ti -1c
.RI "static constexpr bool \fBcheckMName\fP (const short entryNameIndex, const short filterNameIndex) noexcept"
.br
.RI "Checks if the passed entry's and filter's module name indices match Moved from a lambda, so it's easier to use with other parts of the log parser\&. "
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "\fBAEFileReader\fP \fBm_frLogReader\fP"
.br
.RI "The file reader of the opened log file\&. "
.ti -1c
.RI "std::vector< \fBAELogEntryInfo\fP > \fBm_vecEntryIndices\fP"
.br
.RI "The list of all indexed \fIvalid\fP entries in the log file\&. "
.ti -1c
.RI "std::vector< \fBllint\fP > \fBm_vecInvalidEntryIndices\fP"
.br
.RI "The list of all indexed \fIinvalid\fP entries in the log file Each item contains their corresponding cursor position in the file\&. "
.ti -1c
.RI "std::unordered_map< std::string, std::pair< \fBullint\fP, short > > \fBm_mapModuleNames\fP"
.br
.RI "The map of the all module names parsed in the log file\&. "
.ti -1c
.RI "std::array< \fBullint\fP, 9 > \fBm_arrEntryAmount\fP"
.br
.RI "The amount of log entries read in the file, separated by type/severity\&. "
.ti -1c
.RI "std::atomic< std::size_t > \fBm_ullCurrentEntry\fP"
.br
.RI "The number corresponding to the currently-read \fIvalid\fP entry in the log file\&. "
.in -1c
.SH "Detailed Description"
.PP 
ArtyK's Engine's Log Parser; it parses the log files that \fBAELogger\fP writes\&. 

Wrapper around \fBAELogEntry\fP for parsing and \fBAEFileWriter\fP for reading functionality, (ab)using them both\&.
.PP
When opening a file, it reads it and indexes it for log information, and after the parsing/reading the log file to memory can be requested (one entry at a time/the whole thing)\&. Such info would be the amount of entries, amount of entries of each type, etc\&. Also it allows to filter the log by severity and read only important (to the use-case) data\&.
.PP
Hungarian notation is lp\&. (m_lpMyLogParser) Flags start with AELP_ 
.PP
Definition at line \fB66\fP of file \fBAELogParser\&.hpp\fP\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "AELogParser::AELogParser (void)\fR [inline]\fP"

.PP
Class constructor -- just initialises the object\&. 
.PP
Definition at line \fB73\fP of file \fBAELogParser\&.hpp\fP\&.
.SS "AELogParser::AELogParser (const std::string_view fname)\fR [inline]\fP, \fR [explicit]\fP"

.PP
Class constructor -- opens the file and start the indexing process\&. 
.PP
\fBParameters\fP
.RS 4
\fIfname\fP The name of the file to open
.RE
.PP

.PP
Definition at line \fB81\fP of file \fBAELogParser\&.hpp\fP\&.
.SS "AELogParser::~AELogParser (void)\fR [inline]\fP"

.PP
Class destructor -- closes the log file\&. 
.PP
Definition at line \fB90\fP of file \fBAELogParser\&.hpp\fP\&.
.PP
References \fBcloseLog()\fP\&.
.SH "Member Function Documentation"
.PP 
.SS "\fBcint\fP AELogParser::openLog (const std::string_view fname)"

.PP
Opens the log file, reading and parsing every single entry, classifying it by type\&. 
.PP
\fBParameters\fP
.RS 4
\fIfname\fP The name of the file to open
.RE
.PP
\fBReturns\fP
.RS 4
AELP_ERR_NOERROR (0) on success, or AEFR_ERR_* (-1 to -8) or AELE_ERR_* (-11 to -15) flags on error
.RE
.PP

.PP
Definition at line \fB11\fP of file \fBAELogParser\&.cpp\fP\&.
.PP
References \fBAEFR_ERR_NOERROR\fP, \fBAEFR_ERR_OPEN_FILE_ALREADY_OPENED\fP, \fBAELE_FORMAT_MAX_SIZE\fP, \fBAELE_PARSE_STRING_MNAME\fP, \fBAELE_PARSE_STRING_TYPE\fP, \fBAELOG_DEFAULT_QUEUE_SIZE\fP, \fBAELP_ERR_NOERROR\fP, \fBAEFileReader::getCursorPos()\fP, \fBisOpen()\fP, \fBm_arrEntryAmount\fP, \fBm_frLogReader\fP, \fBm_mapModuleNames\fP, \fBm_vecEntryIndices\fP, \fBm_vecInvalidEntryIndices\fP, \fBAEFileReader::openFile()\fP, \fBAELogEntry::parseStringEntry()\fP, and \fBAEFileReader::readStringNL()\fP\&.
.SS "\fBcint\fP AELogParser::closeLog (void)\fR [inline]\fP"

.PP
Closes the log file and clears the memory\&. 
.PP
\fBReturns\fP
.RS 4
return value of the \fBAEFileReader::closefile()\fP (AEFR_ERR_NOERROR if file was closed successfully; AEFR_ERR_FILE_NOT_OPEN if file isn't open)
.RE
.PP

.PP
Definition at line \fB105\fP of file \fBAELogParser\&.hpp\fP\&.
.PP
References \fBAENULL\fP, \fBAEFileReader::closeFile()\fP, \fBm_arrEntryAmount\fP, \fBm_frLogReader\fP, \fBm_mapModuleNames\fP, \fBm_ullCurrentEntry\fP, \fBm_vecEntryIndices\fP, and \fBm_vecInvalidEntryIndices\fP\&.
.SS "\fBcint\fP AELogParser::nextEntry (\fBAELogEntry\fP & entry, const \fBcint\fP severity = \fR\fBAELOG_TYPE_DEBUG\fP\fP, const bool strictSeverity = \fRfalse\fP, const std::string_view mname = \fR''\fP)"

.PP
Read the next \fIvalid\fP entry in the log file of the given severity and module name filter, and parse it to the given \fBAELogEntry\fP object\&. 
.PP
\fBNote\fP
.RS 4
If the strictSeverity is false, then the severity value just changes the lowest limit of the log severity\&. Otherwise it sets the exact severity to look for 
.PP
The module name filter is applied after the severity filter\&. 
.PP
AELOG_TYPE_INVALID works the same as AELOG_TYPE_DEBUG\&. This function parses only \fIvalid\fP entries\&.
.RE
.PP
\fBParameters\fP
.RS 4
\fIentry\fP The log entry object to parse things into
.br
\fIseverity\fP The severity of the log entry to look for
.br
\fIstrictSeverity\fP The flag to indicate whether the search for severity should be strict (exact)
.br
\fImname\fP The module name of the log entry to search for
.RE
.PP
\fBReturns\fP
.RS 4
AELP_ERR_NOERROR (0) on success, or AEFR_ERR_* (-1 to -8) or AELE_ERR_* (-11 to -15) flags on error; error codes from \fBAELogParser::errorFromAELEI()\fP
.RE
.PP
\fBSee also\fP
.RS 4
\fBAELogParser::errorFromAELEI()\fP 
.RE
.PP

.PP
Definition at line \fB75\fP of file \fBAELogParser\&.cpp\fP\&.
.PP
References \fBAEFR_ERR_NOERROR\fP, \fBAEFR_ERR_READ_EOF\fP, \fBAELE_FORMAT_MAX_SIZE\fP, \fBAELE_PARSE_STRING_FULL\fP, \fBAELP_ERR_NOERROR\fP, \fBm_frLogReader\fP, \fBnextEntryCursor()\fP, \fBAELogEntry::parseStringEntry()\fP, \fBAEFileReader::readStringNL()\fP, and \fBAEFileReader::setCursorPos()\fP\&.
.SS "\fBcint\fP AELogParser::nextEntryName (\fBAELogEntry\fP & entry, const std::string_view mname)\fR [inline]\fP"

.PP
Read the next \fIvalid\fP entry in the log file of the given module name, and parse it into the given \fBAELogEntry\fP object\&. 
.PP
\fBSee also\fP
.RS 4
\fBAELogParser::nextEntry()\fP
.RE
.PP
\fBParameters\fP
.RS 4
\fIentry\fP The log entry object to parse things into
.br
\fImname\fP The module name of the log entry to search for
.RE
.PP
\fBReturns\fP
.RS 4
AELP_ERR_NOERROR (0) on success, or AEFR_ERR_* (-1 to -8) or AELE_ERR_* (-11 to -15) flags on error; error codes from \fBAELogParser::errorFromAELEI()\fP
.RE
.PP
\fBSee also\fP
.RS 4
\fBAELogParser::errorFromAELEI()\fP 
.RE
.PP

.PP
Definition at line \fB142\fP of file \fBAELogParser\&.hpp\fP\&.
.PP
References \fBAELP_SEVERITY_ALL\fP, and \fBnextEntry()\fP\&.
.SS "\fBcint\fP AELogParser::nextEntryType (\fBAELogEntry\fP & entry, const \fBcint\fP severity, const bool strictSeverity = \fRfalse\fP)\fR [inline]\fP"

.PP
Read the next \fIvalid\fP entry in the log file of the given severity, and parse it to the given \fBAELogEntry\fP object\&. 
.PP
\fBSee also\fP
.RS 4
AELogEntry::nextEntry()
.RE
.PP
\fBParameters\fP
.RS 4
\fIentry\fP The log entry object to parse things into
.br
\fIseverity\fP The lowest limit of severity of the log entry to look for
.br
\fIstrictSeverity\fP The flag to indicate whether the search for severity should be strict (exact)
.RE
.PP
\fBReturns\fP
.RS 4
AELP_ERR_NOERROR (0) on success, or AEFR_ERR_* (-1 to -8) or AELE_ERR_* (-11 to -15) flags on error; error codes from \fBAELogParser::errorFromAELEI()\fP
.RE
.PP
\fBSee also\fP
.RS 4
\fBAELogParser::errorFromAELEI()\fP 
.RE
.PP

.PP
Definition at line \fB153\fP of file \fBAELogParser\&.hpp\fP\&.
.PP
References \fBAELP_NO_MODULENAME\fP, and \fBnextEntry()\fP\&.
.SS "\fBllint\fP AELogParser::nextEntryCursor (const \fBcint\fP severity = \fR\fBAELOG_TYPE_DEBUG\fP\fP, const bool strictSeverity = \fRfalse\fP, const std::string_view mname = \fR''\fP)\fR [inline]\fP"

.PP
Read the next indexed \fIvalid\fP log entry in the file of the given severity and module name filter, and return it's file cursors in the file\&. 
.PP
\fBNote\fP
.RS 4
The severity value just changes the lowest limit of the log severity (lowest by default is debug)\&. If a higher severity is encountered, it's read as well\&. 
.PP
AELOG_TYPE_INVALID works the same as AELOG_TYPE_DEBUG\&. This function parses only \fIvalid\fP entries\&.
.RE
.PP
\fBParameters\fP
.RS 4
\fIseverity\fP The lowest severity of the log to find
.br
\fIstrictSeverity\fP The flag to indicate whether the search for severity should be strict (exact)
.br
\fImname\fP The module name of the log entry to search for
.RE
.PP
\fBReturns\fP
.RS 4
The file cursors of the next valid entry (in the currently-opened log file); error codes from \fBAELogParser::errorFromAELEI()\fP
.RE
.PP
\fBSee also\fP
.RS 4
\fBAELogParser::errorFromAELEI()\fP 
.RE
.PP

.PP
Definition at line \fB167\fP of file \fBAELogParser\&.hpp\fP\&.
.PP
References \fBAELP_ERR_NOERROR\fP, \fBAELogEntryInfo::cursorIndex\fP, \fBerrorFromAELEI()\fP, and \fBfindNextEntry()\fP\&.
.SS "\fBllint\fP AELogParser::nextEntryCursorName (const std::string_view mname)\fR [inline]\fP"

.PP
Read the next indexed \fIvalid\fP log entry in the file of the given module name filter, and return it's file cursors in the file\&. 
.PP
\fBSee also\fP
.RS 4
\fBAELogParser::nextEntryCursor()\fP
.RE
.PP
\fBParameters\fP
.RS 4
\fImname\fP The module name of the log entry to search for
.RE
.PP
\fBReturns\fP
.RS 4
The file cursors of the next valid entry (in the currently-opened log file); error codes from \fBAELogParser::errorFromAELEI()\fP
.RE
.PP
\fBSee also\fP
.RS 4
\fBAELogParser::errorFromAELEI()\fP 
.RE
.PP

.PP
Definition at line \fB185\fP of file \fBAELogParser\&.hpp\fP\&.
.PP
References \fBAELP_SEVERITY_ALL\fP, and \fBnextEntryCursor()\fP\&.
.SS "\fBllint\fP AELogParser::nextEntryCursorType (const \fBcint\fP severity, const bool strictSeverity = \fRfalse\fP)\fR [inline]\fP"

.PP
Read the next indexed \fIvalid\fP log entry in the file of the given severity filter and return it's file cursors in the file\&. 
.PP
\fBNote\fP
.RS 4
AELOG_TYPE_INVALID works the same as AELOG_TYPE_DEBUG\&. This function parses only \fIvalid\fP entries\&. 
.RE
.PP
\fBSee also\fP
.RS 4
\fBAELogParser::nextEntryCursor()\fP
.RE
.PP
\fBParameters\fP
.RS 4
\fIseverity\fP The lowest severity of the log to find
.br
\fIstrictSeverity\fP The flag to indicate whether the search for severity should be strict (exact)
.RE
.PP
\fBReturns\fP
.RS 4
The file cursors of the next valid entry (in the currently-opened log file); error codes from \fBAELogParser::errorFromAELEI()\fP
.RE
.PP
\fBSee also\fP
.RS 4

AELogParser/src/AELogParser.cpp

/*! @file AELogParser/include/AELogParser.hpp
 *  This file contains the definitions of the engine's [stuff]
 *
 *  Should not cause everything to break :)
 */

#include "../include/AELogParser.hpp"
#include <functional>

cint AELogParser::openLog(const std::string_view fname) {
	if (this->isOpen()) {
		return AEFR_ERR_OPEN_FILE_ALREADY_OPENED;
	}
	const cint ret = this->m_frLogReader.openFile(fname);
	if (ret != AELP_ERR_NOERROR) {
		return ret;
	}

	// and index/parse log
	char fileline[AELE_FORMAT_MAX_SIZE + 1]{}; //+1 more character to determine the validity)
	cint fileRead = AEFR_ERR_NOERROR;
	cint parsedEntry = AELP_ERR_NOERROR;
	llint cursor = 0;
	AELogEntry dummyentry{};
	short mnameIndex = 0;


	m_vecEntryIndices.reserve(AELOG_DEFAULT_QUEUE_SIZE * 10);
	m_vecInvalidEntryIndices.reserve(AELOG_DEFAULT_QUEUE_SIZE);
	m_mapModuleNames.reserve(AELOG_DEFAULT_QUEUE_SIZE);

	while (1) {
		cursor = this->m_frLogReader.getCursorPos();
		fileRead = this->m_frLogReader.readStringNL(fileline, sizeof(fileline) - 1);

		if (fileline[0] == '\n') { // it's an empty line -- skipping...
			break;
		}
		parsedEntry = AELogEntry::parseStringEntry(dummyentry, fileline, AELE_PARSE_STRING_TYPE | AELE_PARSE_STRING_MNAME);


		switch (parsedEntry) {
			case AELP_ERR_NOERROR:
				if (this->m_mapModuleNames.contains(dummyentry.m_sModuleName)) {
					this->m_vecEntryIndices.emplace_back(cursor, this->m_mapModuleNames[dummyentry.m_sModuleName].second, dummyentry.m_cLogType);
				}
				else {
					this->m_mapModuleNames[dummyentry.m_sModuleName].second = ++mnameIndex;
					this->m_vecEntryIndices.emplace_back(cursor, mnameIndex, dummyentry.m_cLogType); // a bit of optimisation
				}
				
				this->m_mapModuleNames[dummyentry.m_sModuleName].first++;
				this->m_arrEntryAmount[dummyentry.m_cLogType + 1]++;
				break;

			default:
				if (fileRead != AEFR_ERR_NOERROR) {
					return ret;
				}
				this->m_vecInvalidEntryIndices.push_back(cursor);
				this->m_arrEntryAmount[0]++;
				break;
		}

	}

	m_vecEntryIndices.shrink_to_fit();
	m_vecInvalidEntryIndices.shrink_to_fit();

	return ret;
}


cint AELogParser::nextEntry(AELogEntry& entry, const cint severity, const bool strictSeverity, const std::string_view mname) {

// 	const AELogEntryInfo leInfo = this->findNextEntry(severity, mname, strictSeverity);
// 	
// 	cint readret = AELogParser::errorFromAELEI(leInfo);
// 	if (readret != AELP_ERR_NOERROR) {
// 		return readret; // return the error code from the AELEI
// 	}
	
	//we implemented this already lol
	llint readret = this->nextEntryCursor(severity, strictSeverity, mname);
	if (readret != AELP_ERR_NOERROR) {
		return readret; // return the error
	}


	char str[AELE_FORMAT_MAX_SIZE + 2]{}; // 1 character more than the log entry - to determine the validity with size

	this->m_frLogReader.setCursorPos(readret, SEEK_SET);

	readret = this->m_frLogReader.readStringNL(str, sizeof(str) - 1);
	// check even if it's EOF. It's not an error, but a warning from file reader
	// that we have reached the end, and *may* have not read everything that we wanted
	// The real result (if we actually didn't read what we wanted) will be given from the AELogEntry::parseStringEntry()
	if (readret == AEFR_ERR_NOERROR || readret == AEFR_ERR_READ_EOF) { 
		return AELogEntry::parseStringEntry(entry, str, AELE_PARSE_STRING_FULL);
	}
	
	return readret; // error I guess
}


cint AELogParser::logToQueue(AELogEntry*& begin, const cint severity, const bool strictSeverity, const std::string_view mname) {
	_AELP_CHECK_IF_FILE_OPEN;
	
	AELogEntry* ptr = begin = AELogEntry::makeQueue((strictSeverity) ? this->amountEntriesType(severity) : this->amountEntriesValid(severity), false, nullptr);
	cint retval = 0;
	// Yes, this is..weird, But in nextEntry it will pre-increment it before checking
	// whether it's EOF or not. So...set it to 1 less from the start, instead of subtracting 1 every time
	// Just an optimisation, uint over/underflow is well-defined anyway
	const std::size_t entrynum = this->m_ullCurrentEntry;
	this->m_ullCurrentEntry = -1; 
	
	int test = 0;

	while (ptr && (retval = this->nextEntry(*ptr, severity, strictSeverity, AELP_NO_MODULENAME)) == AELP_ERR_NOERROR) {
		ptr = ptr->m_pNextNode;
		test++;
	}

	this->m_ullCurrentEntry = entrynum;

	if (mname.size()) { // we have module name passed
		retval = AELogParser::filterQueueName(begin, mname);
	}

	return retval;
}

cint AELogParser::filterQueue(AELogEntry*& ptr, const cint severity, const bool strictSeverity, const std::string_view mname, const bool reallocQueue) {
	// error handling
	if (ptr == nullptr) { // nullptr was passed, lol
		return AELP_ERR_INVALID_QUEUE;
	}
	// its all the sameeeee
	if (severity == AELP_SEVERITY_ALL && mname.empty()) {
		return AELP_ERR_INVALID_FILTER; // invalid filter option
	}

	AELogEntry* cur = ptr;
	AELogEntry kludge{ .m_pNextNode = ptr }; // a kludge variable for the pastEntry to work
	AELogEntry* past = &kludge;
	std::size_t newQueueSize = 0;

	const std::size_t szCheck = ace::math::min<std::size_t>(mname.size(), AELE_MODULENAME_SIZE);
	const auto cmpMName = (mname.empty()) ?
		([](const char* const entryName, const std::size_t szCheck, const std::string_view& filterMname) noexcept { return true; }) : // if the module name is empty -- return true always (we aren't checking for it)
		([](const char* const entryName, const std::size_t szCheck, const std::string_view& filterMname) noexcept { return !std::memcmp(entryName, filterMname.data(), szCheck); });


	//check if it even has the stuff matching the filter
	while (cur) {
		if (AELogParser::checkSeverity(cur->m_cLogType, severity, strictSeverity) && cmpMName(cur->m_sModuleName, szCheck, mname)) {
			goto foundSeverity; // we found the severity, break the loop
			break;
		}

		cur = cur->m_pNextNode;
	}

	//didn't quite catch it: error reporting
	// there is no log entry with the given severity
	if (mname.empty()) {
		return AELP_ERR_INVALID_SEVERITY; // no module name filter
	}
	if (severity == AELP_SEVERITY_ALL) {
		return AELP_ERR_INVALID_MODULE_NAME; // no severity filter
	}
	return AELP_ERR_FILTER_NO_MATCHES; //both exist, but none were captured -- no matches!

	//welp, we found it
foundSeverity:
	while (cur) {
		if (AELogParser::checkSeverity(cur->m_cLogType, severity, strictSeverity) && cmpMName(cur->m_sModuleName, szCheck, mname)) {
			past = past->m_pNextNode;
			past->copyEntry(*cur);
			newQueueSize++;
		}
		cur = cur->m_pNextNode;

	}
	past->m_pNextNode = nullptr;

	if (reallocQueue) {
		//clean up the queue
		//or...make a new, smaller queue instead!
		cur = ptr;
		AELogEntry* const newQueue = AELogEntry::makeQueue(newQueueSize, false);
		AELogEntry* iter = newQueue;
		while (cur) {
			iter->copyEntry(*cur);
			iter = iter->m_pNextNode;
			cur = cur->m_pNextNode;
		}
		delete[] ptr;
		ptr = newQueue;
	}

	return AELP_ERR_NOERROR;
}

AELogEntryInfo AELogParser::findNextEntry(const cint severity, const std::string_view mname, const bool strictSeverity) {

	if (this->isClosed()) {
		return AELogEntryInfo::invalidEntry();
	}

	std::string mnameStr = mname.data();

	if (!mname.empty() && !this->m_mapModuleNames.contains(mnameStr)) {
		return AELogEntryInfo{ .mnameIndex = AELEI_INVLAID_MNAME };
	}

	if (!ace::utils::isInRange(AELP_SEVERITY_ALL, AELOG_TYPE_FATAL_ERROR, severity) || //is in range at all?
		!this->containsSeverity(severity, strictSeverity)) { // does this exist in the log (accounting for strict severity filter)
		return AELogEntryInfo{ .logType = AELEI_INVALID_TYPE };
	}

	
	const short filterMnameIndex = (!mname.empty()) ? this->m_mapModuleNames[mnameStr].second : 0;

	mnameStr.clear();
	mnameStr.shrink_to_fit(); //not necessary anymore

	const std::size_t entryNum = m_ullCurrentEntry; //write down the cursor value

	//searching time
	while (1) {

		if (++this->m_ullCurrentEntry >= this->m_vecEntryIndices.size()) {
			// check if cursor was already out of bounds
			if ((entryNum + 1) >= this->m_vecEntryIndices.size()) {
				return AELogEntryInfo{ .cursorIndex = AELEI_INVALID_CURSOR };
			}
			//nope, we got out of bounds ourselves
			//because of invalid filter!
			
			this->m_ullCurrentEntry = entryNum; // return the cursor to the previous position
			return AELogEntryInfo { 
				//god damn it msvc. Why is it an error to do it *that* way, while initialising it with a number outside of ternary is okay
				.mnameIndex = (short) ((filterMnameIndex) ? AELEI_INVLAID_MNAME : 0), 
				.logType = (cint) ((severity != AELP_SEVERITY_ALL) ? AELEI_INVALID_TYPE : 0)
			};

		}
		if (AELogParser::checkSeverity(this->m_vecEntryIndices[this->m_ullCurrentEntry].logType, severity, strictSeverity) && //check the severity with lambdas
			AELogParser::checkMName(this->m_vecEntryIndices[this->m_ullCurrentEntry].mnameIndex, filterMnameIndex)) {

			break;
		}
	}

	return this->m_vecEntryIndices[this->m_ullCurrentEntry];

}